modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -4,129 +4,192 @@ import cats.data.{EitherT, NonEmptyList, NonEmptySet}
+        if (isSigned) {
+          parseSignedTransaction(hex)
+        } else {
+          parseUnsignedTransaction(hex)
         }
 
-  }
+      def combineTransaction(hex: Hex, signature: RosettaSignature): EitherT[F, ConstructionError, Hex] =
+        EitherT
+          .fromEither(KryoSerializer[F].deserialize[DAGTransaction](hex.toBytes))
+          .leftMap(_ => MalformedTransaction)
+          .flatMap { transaction =>
+            EitherT {
+              signature.publicKey.hexBytes.toPublicKeyByEC
+                .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
+            }.flatMap { proof =>
+              EitherT
+                .fromEither(
+                  KryoSerializer[F]
+                    .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
+                )
+                .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
+                .map(Hex.fromBytes(_))
+            }
+          }
+
+      override def getPayloads(
+        operations: NonEmptyList[Operation],
+        metadata: TransactionMetadata
+      ): EitherT[F, ConstructionError, PayloadsResult] =
+        EitherT(
+          saltGen().map(transactionSalt =>
+            for {
+              (positiveOperation, negativeOperation) <- Either.fromOption(
+                getPayloadOperations(operations),
+                InvalidPayloadOperations
+              )
+
+              _ <- Either.cond(
+                positiveOperation.amount.negate.toOption === negativeOperation.amount.some,
+                (),
+                InvalidPayloadOperations
+              )
+
+              transactionAmount <- positiveOperation.amount.value.toTransactionAmount
+                .leftMap(_ => InvalidPayloadOperations)
+
+              sourceAddress = negativeOperation.account.address
+              transactionFee <-
+                metadata.suggestedFee match {
+                  case None      => TransactionFee.zero.asRight
+                  case Some(amt) => amt.value.toTransactionFee.leftMap(_ => InvalidSuggestedFee)
+                }
+
+              dagTransaction =
+                DAGTransaction(
+                  source = sourceAddress,
+                  destination = positiveOperation.account.address,
+                  amount = transactionAmount,
+                  fee = transactionFee,
+                  parent = metadata.lastReference,
+                  salt = transactionSalt
+                )
+
+              serializedTxn <-
+                KryoSerializer[F]
+                  .serialize(dagTransaction)
+                  .leftMap(e => SerializationError(e.getMessage))
+
+              txHash <-
+                serializedTxn.hash
+                  .leftMap[ConstructionError](_ => MalformedTransaction)
+                  .map(h => Hex(h.value))
+
+              payload = SigningPayload(AccountIdentifier(sourceAddress, none), txHash, SignatureType.ECDSA)
```

Comment:
I think it's wrong to pass `txHash` here.
`SigningPayload` has `hexBytes` property which rosetta specifies as **_Hex-encoded string of the payload bytes._**. You don't sign the hash but the transaction, so imo you should pass here the transaction data (exactly the same what we pass to sign the transaction out of rosetta).

Reply:
@sdavidp It's the specification for exactly the endpoint you just implementing
https://www.rosetta-api.org/docs/models/ConstructionPayloadsResponse.html (https://www.rosetta-api.org/docs/models/SigningPayload.html)

Reply:
@sdavidp why have you resolved the conversation? I see neither the fix nor the msg that my comment is wrong 🤔

Reply:
@marcinwadon Based on the comment by @TheMMaciek I changed the code to take the hash of the DAGTransaction, not it's serialized bytes.  If that's still not correct can you please clarify what's wrong?

Reply:
Which comment you're referring to? I had to miss that

Reply:
Anyway, IMO it should be the same flow as singing the transaction out of Rosetta, so you should sign the transaction data, not a single hash 🤔

Reply:
@marcinwadon Ok, with @ryle-goehausen 's help I've tweaked this again. Does this look correct?